src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,132 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+        size = 0;
+    }
 
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (!isIndexValid(index) && index != size) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        if (index == size) {
+            add(value);
+        } else {
+            addElementInMiddle(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        while (list.size() + size > elementData.length) {
+            elementData = grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (!isIndexValid(index)) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (!isIndexValid(index)) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (!isIndexValid(index)) {
+            throw new ArrayIndexOutOfBoundsException();
+        }
+        T removedElement = get(index);
+        arrayShift(index);
+        return removedElement;
     }
 
     @Override
     public T remove(T t) {
-        return null;
+        if (!isIndexValid(getIndex(t))) {
```

Comment:
let's save the result of getIndex() to local variable in order not to invocate it twice